File: python/tests/test_companion_types.py
"""test_companion_types.py

Validates:
 - InterfaceKind enum values
 - CompanionContext, CompanionTurn, CompanionProactiveEvent construction
 - FaceAffectMapper.apply() mutations against facex_biometric_vectors.json
 - FaceCompanionBridge.observe() proactive event logic
"""
from __future__ import annotations

import json
import pathlib
from datetime import datetime, timezone

import pytest

from circle_ai.companion.companion_types import (
    CompanionContext,
    CompanionProactiveEvent,
    CompanionTurn,
    InterfaceKind,
)
from circle_ai.companion import face_affect_mapper, face_companion_bridge
from circle_ai.memory.affect_state import AffectState
from circle_ai.tools.facial_metric_matrix import (
    FaceBoundingBox,
    FaceExpressionClassification,
    FacialMetricMatrix,
)

FIXTURES_DIR = pathlib.Path(__file__).parent.parent.parent / "fixtures"
EPSILON = 1e-5


def _utc() -> datetime:
    return datetime.now(timezone.utc)


def _load_affect_mapper_vectors() -> list[dict]:
    with open(FIXTURES_DIR / "facex_biometric_vectors.json", encoding="utf-8") as f:
        return json.load(f)["affect_mapper_vectors"]


AFFECT_MAPPER_VECTORS = _load_affect_mapper_vectors()

_EXPRESSION_MAP: dict[str, FaceExpressionClassification] = {
    "Happy":     FaceExpressionClassification.HAPPY,
    "Surprised": FaceExpressionClassification.SURPRISED,
    "Confused":  FaceExpressionClassification.CONFUSED,
    "Stressed":  FaceExpressionClassification.STRESSED,
    "Angry":     FaceExpressionClassification.ANGRY,
    "Neutral":   FaceExpressionClassification.NEUTRAL,
    "Unknown":   FaceExpressionClassification.UNKNOWN,
}


def _make_matrix(expression: str, confidence: float) -> FacialMetricMatrix:
    return FacialMetricMatrix(
        landmarks=[0.0] * 136,
        bounding_box=FaceBoundingBox(x=0.0, y=0.0, width=1.0, height=1.0),
        expression=_EXPRESSION_MAP[expression],
        confidence_score=confidence,
    )


def _make_affect(d: dict) -> AffectState:
    s = AffectState()
    s.curiosity   = float(d["curiosity"])
    s.engagement  = float(d["engagement"])
    s.uncertainty = float(d["uncertainty"])
    s.rapport     = float(d["rapport"])
    s.energy      = float(d["energy"])
    return s


# ---------------------------------------------------------------------------
# InterfaceKind
# ---------------------------------------------------------------------------

def test_interface_kind_count() -> None:
    assert len(InterfaceKind) == 7


def test_interface_kind_values() -> None:
    expected = {"Mobile", "Wearable", "Desktop", "Web", "IoT", "Ambient", "Headless"}
    actual   = {member.value for member in InterfaceKind}
    assert actual == expected


# ---------------------------------------------------------------------------
# CompanionContext round-trip
# ---------------------------------------------------------------------------

def test_companion_context_fields() -> None:
    now = _utc()
    ctx = CompanionContext(
        identity_id="id-001",
        display_name="Sipho",
        preferred_language="zu",
        interface=InterfaceKind.MOBILE,
        persona_hints="[User preferences]\nKeep responses brief.\n",
        affect_summary="[Affect state]\nYou are fully engaged.\n",
        recent_memory_snippets=["snippet-1", "snippet-2"],
        active_goals=["Finish project"],
        context_built_at=now,
    )
    assert ctx.identity_id == "id-001"
    assert ctx.display_name == "Sipho"
    assert ctx.preferred_language == "zu"
    assert ctx.interface is InterfaceKind.MOBILE
    assert ctx.context_built_at == now


def test_companion_context_optional_language() -> None:
    ctx = CompanionContext(
        identity_id="anon",
        display_name="Guest",
        preferred_language=None,
        interface=InterfaceKind.IOT,
        persona_hints="",
        affect_summary="",
        recent_memory_snippets=[],
        active_goals=[],
        context_built_at=_utc(),
    )
    assert ctx.preferred_language is None


def test_companion_context_is_frozen() -> None:
    ctx = CompanionContext(
        identity_id="x",
        display_name="X",
        preferred_language=None,
        interface=InterfaceKind.HEADLESS,
        persona_hints="",
        affect_summary="",
        recent_memory_snippets=[],
        active_goals=[],
        context_built_at=_utc(),
    )
    with pytest.raises((AttributeError, TypeError)):
        ctx.display_name = "Y"  # type: ignore[misc]


# ---------------------------------------------------------------------------
# CompanionTurn round-trip
# ---------------------------------------------------------------------------

def test_companion_turn_user() -> None:
    now = _utc()
    turn = CompanionTurn(role="user", content="Hello, B!", timestamp=now)
    assert turn.role == "user"
    assert turn.content == "Hello, B!"
    assert turn.timestamp == now


def test_companion_turn_is_frozen() -> None:
    turn = CompanionTurn(role="user", content="hi", timestamp=_utc())
    with pytest.raises((AttributeError, TypeError)):
        turn.content = "changed"  # type: ignore[misc]


# ---------------------------------------------------------------------------
# CompanionProactiveEvent round-trip
# ---------------------------------------------------------------------------

def test_companion_proactive_event() -> None:
    now = _utc()
    event = CompanionProactiveEvent(
        session_id="sess-42",
        identity_id="id-001",
        interface=InterfaceKind.WEARABLE,
        message="Time to stretch.",
        trigger_name="posture_reminder",
        generated_at=now,
    )
    assert event.session_id == "sess-42"
    assert event.identity_id == "id-001"
    assert event.interface is InterfaceKind.WEARABLE
    assert event.trigger_name == "posture_reminder"
    assert event.generated_at == now


# ---------------------------------------------------------------------------
# FaceAffectMapper — parametrised against fixtures
# ---------------------------------------------------------------------------

@pytest.mark.parametrize("entry", AFFECT_MAPPER_VECTORS, ids=[e["id"] for e in AFFECT_MAPPER_VECTORS])
def test_face_affect_mapper(entry: dict) -> None:
    affect = _make_affect(entry["initial_affect"])
    matrix = _make_matrix(entry["expression"], float(entry["confidence"]))
    face_affect_mapper.apply(matrix, affect)

    exp = entry["expected_affect"]
    tol = float(entry.get("tolerance", EPSILON))

    assert abs(affect.curiosity   - float(exp["curiosity"]))   <= tol, \
        f"[{entry['id']}] curiosity mismatch"
    assert abs(affect.engagement  - float(exp["engagement"]))  <= tol, \
        f"[{entry['id']}] engagement mismatch"
    assert abs(affect.uncertainty - float(exp["uncertainty"])) <= tol, \
        f"[{entry['id']}] uncertainty mismatch"
    assert abs(affect.rapport     - float(exp["rapport"]))     <= tol, \
        f"[{entry['id']}] rapport mismatch"
    assert abs(affect.energy      - float(exp["energy"]))      <= tol, \
        f"[{entry['id']}] energy mismatch"


# ---------------------------------------------------------------------------
# FaceCompanionBridge.observe() — proactive event triggering
# ---------------------------------------------------------------------------

def test_observe_confused_high_uncertainty_triggers_event() -> None:
    """Confused expression with high confidence from uncertain state should produce an event."""
    affect = AffectState()
    affect.uncertainty = 0.75  # above CONFUSION_THRESHOLD

    matrix = _make_matrix("Confused", 0.79)
    # After applying the Confused expression, uncertainty will increase further
    # Reset to a state that will be >= threshold after apply
    affect.uncertainty = 0.70  # exactly at threshold

    event = face_companion_bridge.observe(
        matrix, affect, "sess-1", "id-001", InterfaceKind.MOBILE
    )
    assert event is not None
    assert event.trigger_name == "face_confusion_detected"
    assert event.session_id == "sess-1"
    assert event.identity_id == "id-001"
    assert event.interface is InterfaceKind.MOBILE


def test_observe_stressed_high_uncertainty_triggers_event() -> None:
    """Stressed expression with high confidence from uncertain state should produce an event."""
    affect = AffectState()
    affect.uncertainty = 0.65  # after +0.08 from stressed = 0.73 >= 0.70

    matrix = _make_matrix("Stressed", 0.85)
    event = face_companion_bridge.observe(
        matrix, affect, "sess-2", "id-002", InterfaceKind.DESKTOP
    )
    assert event is not None
    assert event.trigger_name == "face_confusion_detected"


def test_observe_neutral_no_event() -> None:
    """Neutral expression must never produce a proactive event."""
    affect = AffectState()
    affect.uncertainty = 0.9  # very uncertain, but neutral expression

    matrix = _make_matrix("Neutral", 0.95)
    event = face_companion_bridge.observe(
        matrix, affect, "sess-3", "id-003", InterfaceKind.WEB
    )
    assert event is None


def test_observe_happy_no_event() -> None:
    """Happy expression must not produce a proactive event regardless of uncertainty."""
    affect = AffectState()
    affect.uncertainty = 0.9

    matrix = _make_matrix("Happy", 0.92)
    event = face_companion_bridge.observe(
        matrix, affect, "sess-4", "id-004", InterfaceKind.WEARABLE
    )
    assert event is None


def test_observe_low_confidence_no_event() -> None:
    """Low confidence (< 0.5) must produce no event and no affect mutation."""
    affect = AffectState()
    affect.uncertainty = 0.9

    matrix = _make_matrix("Stressed", 0.49)
    event = face_companion_bridge.observe(
        matrix, affect, "sess-5", "id-005", InterfaceKind.MOBILE
    )
    assert event is None
    # Affect should be unchanged (confidence below threshold)
    assert abs(affect.uncertainty - 0.9) <= EPSILON

File: c/tests/test_companion_types.c
/*
 * test_companion_types.c — Companion types, FaceAffectMapper, and
 *                          FaceCompanionBridge tests.
 *
 * Expression deltas match fixtures/facex_biometric_vectors.json
 * affect_mapper_vectors within epsilon=1e-5.
 * Returns 0 on all-pass, calls assert() on first failure.
 */

#include <stdio.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include "circle_ai/circle_ai.h"

#define EPSILON 1e-5f

static void check(const char *id, float got, float expected) {
    float diff = fabsf(got - expected);
    if (diff > EPSILON) {
        fprintf(stderr, "FAIL [%s]: got %.8f, expected %.8f (diff %.8f)\n",
                id, (double)got, (double)expected, (double)diff);
        assert(0);
    }
}

int main(void) {
    /* ------------------------------------------------------------------
     * Companion struct construction
     * ------------------------------------------------------------------ */
    ca_companion_context_t ctx;
    memset(&ctx, 0, sizeof(ctx));
    strncpy(ctx.session_id,  "550e8400-e29b-41d4-a716-446655440000", 36);
    strncpy(ctx.identity_id, "550e8400-e29b-41d4-a716-446655440001", 36);
    ctx.interface_kind = CA_IFACE_VOICE;
    ctx.locale = "en-US";
    ctx.started_at = 1704067200000LL;

    assert(ctx.interface_kind == CA_IFACE_VOICE);
    assert(strcmp(ctx.locale, "en-US") == 0);
    assert(ctx.started_at == 1704067200000LL);

    /* All interface kinds */
    assert(CA_IFACE_VOICE   == 0);
    assert(CA_IFACE_TEXT    == 1);
    assert(CA_IFACE_VISUAL  == 2);
    assert(CA_IFACE_AMBIENT == 3);

    ca_companion_turn_t turn;
    memset(&turn, 0, sizeof(turn));
    strncpy(turn.turn_id,    "550e8400-e29b-41d4-a716-446655440000", 36);
    strncpy(turn.session_id, "550e8400-e29b-41d4-a716-446655440001", 36);
    turn.user_input          = "Hello";
    turn.assistant_response  = "Hi there!";
    turn.turn_index          = 0;

    assert(turn.turn_index == 0);
    assert(strcmp(turn.user_input, "Hello") == 0);
    assert(strcmp(turn.assistant_response, "Hi there!") == 0);

    /* Proactive event kinds */
    assert(CA_PROACTIVE_IDLE_TOO_LONG   == 0);
    assert(CA_PROACTIVE_TOPIC_SHIFT     == 1);
    assert(CA_PROACTIVE_GOAL_COMPLETED  == 2);
    assert(CA_PROACTIVE_GOAL_SUGGESTED  == 3);
    assert(CA_PROACTIVE_MEMORY_RECALLED == 4);

    ca_proactive_event_t event;
    memset(&event, 0, sizeof(event));
    event.kind    = CA_PROACTIVE_GOAL_SUGGESTED;
    event.payload = "{\"goal\":\"test\"}";
    assert(event.kind == CA_PROACTIVE_GOAL_SUGGESTED);
    assert(strcmp(event.payload, "{\"goal\":\"test\"}") == 0);

    /* ------------------------------------------------------------------
     * FaceAffectMapper — fixture: affect_mapper_vectors
     * ------------------------------------------------------------------ */

    ca_affect_state_t s;
    bool mutated;

    /* happy_from_neutral: engagement += 0.03, energy += 0.02 */
    s = ca_affect_state_default();
    mutated = ca_face_apply_affect(0.92f, CA_FACE_HAPPY, &s);
    assert(mutated == true);
    check("happy engagement", s.engagement, 0.53f);
    check("happy energy",     s.energy,     0.52f);
    check("happy curiosity",  s.curiosity,  0.5f);
    check("happy uncertainty",s.uncertainty,0.2f);
    check("happy rapport",    s.rapport,    0.0f);

    /* surprised_from_neutral: curiosity += 0.04 */
    s = ca_affect_state_default();
    mutated = ca_face_apply_affect(0.88f, CA_FACE_SURPRISED, &s);
    assert(mutated == true);
    check("surprised curiosity",   s.curiosity,   0.54f);
    check("surprised engagement",  s.engagement,  0.5f);
    check("surprised uncertainty", s.uncertainty, 0.2f);

    /* confused_from_neutral: uncertainty += 0.05 */
    s = ca_affect_state_default();
    mutated = ca_face_apply_affect(0.79f, CA_FACE_CONFUSED, &s);
    assert(mutated == true);
    check("confused uncertainty", s.uncertainty, 0.25f);
    check("confused engagement",  s.engagement,  0.5f);
    check("confused curiosity",   s.curiosity,   0.5f);

    /* stressed_from_neutral: uncertainty += 0.08, energy -= 0.05 */
    s = ca_affect_state_default();
    mutated = ca_face_apply_affect(0.85f, CA_FACE_STRESSED, &s);
    assert(mutated == true);
    check("stressed uncertainty", s.uncertainty, 0.28f);
    check("stressed energy",      s.energy,      0.45f);
    check("stressed engagement",  s.engagement,  0.5f);

    /* angry_from_neutral (rapport=0.3): engagement -= 0.04, rapport -= 0.02 */
    s = ca_affect_state_default();
    s.rapport = 0.3f;
    mutated = ca_face_apply_affect(0.91f, CA_FACE_ANGRY, &s);
    assert(mutated == true);
    check("angry engagement", s.engagement, 0.46f);
    check("angry rapport",    s.rapport,    0.28f);
    check("angry uncertainty",s.uncertainty,0.2f);

    /* neutral_expression_no_change */
    s = ca_affect_state_default();
    mutated = ca_face_apply_affect(0.95f, CA_FACE_NEUTRAL, &s);
    assert(mutated == false);
    check("neutral engagement",  s.engagement,  0.5f);
    check("neutral curiosity",   s.curiosity,   0.5f);
    check("neutral uncertainty", s.uncertainty, 0.2f);
    check("neutral rapport",     s.rapport,     0.0f);
    check("neutral energy",      s.energy,      0.5f);

    /* low_confidence_discarded: confidence < 0.5 => no change */
    s = ca_affect_state_default();
    mutated = ca_face_apply_affect(0.49f, CA_FACE_STRESSED, &s);
    assert(mutated == false);
    check("low_conf uncertainty", s.uncertainty, 0.2f);
    check("low_conf energy",      s.energy,      0.5f);

    /* clamp_max_engagement: near-max + happy delta => 1.0 */
    s = ca_affect_state_default();
    s.engagement = 0.99f;
    mutated = ca_face_apply_affect(0.95f, CA_FACE_HAPPY, &s);
    assert(mutated == true);
    check("clamp_max_engagement", s.engagement, 1.0f);
    check("clamp_max_energy",     s.energy,     0.52f);

    /* ------------------------------------------------------------------
     * FaceCompanionBridge — confusion threshold
     * ------------------------------------------------------------------ */

    /* uncertainty well below threshold => no event */
    s = ca_affect_state_default(); /* uncertainty = 0.2 */
    ca_proactive_event_t out;
    memset(&out, 0, sizeof(out));
    int fired = ca_face_observe(0.80f, CA_FACE_HAPPY, &s,
                                "sess-001", "id-001", CA_IFACE_TEXT, &out);
    assert(fired == 0);

    /* Build up uncertainty above threshold (0.70) via multiple confused signals */
    s = ca_affect_state_default();
    s.uncertainty = 0.66f; /* 0.66 + 0.05 = 0.71 >= 0.70 */
    memset(&out, 0, sizeof(out));
    fired = ca_face_observe(0.80f, CA_FACE_CONFUSED, &s,
                            "sess-002", "id-002", CA_IFACE_TEXT, &out);
    assert(fired == 1);
    assert(s.uncertainty >= CA_CONFUSION_THRESHOLD);
    assert(strcmp(out.session_id,  "sess-002") == 0);
    assert(strcmp(out.identity_id, "id-002")   == 0);

    /* Exactly at threshold: 0.70 >= 0.70 fires if a confused signal just pushed it */
    s = ca_affect_state_default();
    s.uncertainty = 0.65f; /* 0.65 + 0.05 = 0.70 exactly */
    memset(&out, 0, sizeof(out));
    fired = ca_face_observe(0.80f, CA_FACE_CONFUSED, &s,
                            "sess-003", "id-003", CA_IFACE_VOICE, &out);
    assert(fired == 1);
    check("threshold_exact", s.uncertainty, CA_CONFUSION_THRESHOLD);

    printf("All companion type tests passed.\n");
    return 0;
}
